src/main/scala/chisel3/util/BitPat.scala

```diff
@@ -170,5 +170,8 @@ sealed class BitPat(val value: BigInt, val mask: BigInt, width: Int) extends Sou
     }
   }.mkString
 
+  import scala.math.Ordered.orderingToOrdered
+  def compare(that: BitPat): Int = (getWidth, value, mask) compare (that.getWidth, that.value, that.mask)
```

Comment:
It makes me slightly nervous to expose the order of `BitPat` as a public API. Are we 100% sure this is the right, canonical way to order `BitPat`? It makes me a little nervous since if it's the wrong ordering, it's kind of awkward to change if users are sorting `BitPat` themselves. 

I'm not saying this is wrong, but could we instead use the `Ordering` typeclass and keep it private for purposes of sorting internally? No other type provided by Chisel provides `Ordered` so it just makes me a little nervous to do it here for the notionally unrelated purpose of making decoder output deterministic.

Reply:
> Are we 100% sure this is the right, canonical way to order BitPat?

this makes the order of BitPat unique. But I can still remove this to make BitPat unordered, and use sortBy in the place where we need this order. I think both are ok. Let you decide ;p

Reply:
Order being unique is important, but not sufficient, I could make the order be the `.hashCode` order but obviously that wouldn't be very intuitive 🙂. I'd rather `Order` not be part of the public API, but yes it should be used internally to sort things. You could do this with `sortBy`, but alternatively you could have a private (or package private) implicit val implementation of https://www.scala-lang.org/api/2.12.13/scala/math/Ordering.html and put this functionality there. Then you could import it and still use `.sorted`.

Reply:
To be more specific, what I'm asking for here is that instead of extending `Ordered`, we provide a private implementation of `Ordering` like so:
```scala
object BitSet {
  ...
  private implicit val bitSetOrder = new Ordering[BitSet] {
    def compare(that: BitPat): Int = (getWidth, value, mask) compare (that.getWidth, that.value, that.mask)
  }
  ...
}
```
Then we can import it and still use `.sort` internally.

Reply:
wow. I didn't think we can do this!
I will refactor it!